csharp/ql/src/experimental/Security Features/JsonWebTokenHandler/JsonWebTokenHandlerLib.qll

```diff
@@ -0,0 +1,315 @@
+ * It will also hold if no exceptions are thrown by the callable
+ */
+predicate callableOnlyThrowsArgumentNullException(Callable c) {
+  forall(ThrowElement thre | c = thre.getEnclosingCallable() |
+    thre.getThrownExceptionType().hasQualifiedName("System.ArgumentNullException")
+  )
+}
+
+/**
+ * A specialization of `CallableAlwaysReturnsTrue` that takes into consideration exceptions being thrown for higher precision.
+ */
+class CallableAlwaysReturnsTrueHigherPrecision extends CallableAlwaysReturnsTrue {
+  CallableAlwaysReturnsTrueHigherPrecision() {
+    callableOnlyThrowsArgumentNullException(this) and
+    (
+      forall(Call call, Callable callable | call.getEnclosingCallable() = this |
+        callable.getACall() = call and
+        callable instanceof CallableAlwaysReturnsTrueHigherPrecision
+      )
+      or
+      exists(AnonymousFunctionExpr le, Call call, CallableAlwaysReturnsTrueHigherPrecision cat |
+        this = le
+      |
+        le.canReturn(call) and
+        cat.getACall() = call
+      )
+      or
+      exists(LambdaExpr le | le = this |
+        le.getBody() instanceof CallableAlwaysReturnsTrueHigherPrecision
+      )
+    )
+  }
+}
+
+/**
+ * A property Write for the `IssuerValidator` property for `Microsoft.IdentityModel.Tokens.TokenValidationParameters`
+ */
+class TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator extends PropertyWrite {
+  TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator() {
+    exists(Property p, Class c |
+      c.hasQualifiedName("Microsoft.IdentityModel.Tokens.TokenValidationParameters")
+    |
+      p.getAnAccess() = this and
+      c.getAProperty() = p and
+      p.hasName("IssuerValidator")
+    )
+  }
+}
+
+/**
+ * A callable that returns a `string` and has a `string` as 1st argument
+ */
+private class CallableReturnsStringAndArg0IsString extends Callable {
+  CallableReturnsStringAndArg0IsString() {
+    this.getReturnType() instanceof StringType and
+    this.getParameter(0).getType() instanceof StringType
+  }
+}
+
+/**
+ * A Callable that always return the 1st argument, both of `string` type
+ */
+class CallableAlwaysReturnsParameter0 extends CallableReturnsStringAndArg0IsString {
+  CallableAlwaysReturnsParameter0() {
+    forall(ReturnStmt rs | rs.getEnclosingCallable() = this |
+      rs.getExpr() = this.getParameter(0).getAnAccess()
+    ) and
+    exists(ReturnStmt rs | rs.getEnclosingCallable() = this)
+    or
+    exists(AnonymousFunctionExpr le, Call call, CallableAlwaysReturnsParameter0 cat | this = le |
+      call = le.getExpressionBody() and
+      cat.getACall() = call
+    )
+    or
+    this.getBody() = this.getParameter(0).getAnAccess()
+  }
+}
+
+/**
+ * A Callable that always return the 1st argument, both of `string` type. Higher precision
+ */
+class CallableAlwaysReturnsParameter0MayThrowExceptions extends CallableReturnsStringAndArg0IsString {
+  CallableAlwaysReturnsParameter0MayThrowExceptions() {
```

Comment:
A similar rewrite as above should be possible.

Reply:
I will fix this last one tomorrow. I have updated the tests for the new changes, so tomorrow when I change this one I will be able to compare the output & make sure I don't introduce any regressions.

Reply:
Fixed